### PR TITLE
fix: API migration to improve performance in resource pool page

### PR DESCRIPTION
**Pull request by keita-determined**
## Description
[ET-104]
Related to #9048

## Test Plan
- Check if `Slots Allocated` bars show the correct data
- Check if topology works 
  - Check if topology doesn't show up if nodes > 1,000 OR slots > 10,000 
  - Check if polling is working when topology is available
- Check if test cases covers edge cases

## Commentary (optional)
Merge this after #9048

## Checklist

- [x] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[ET-104]: https://hpe-aiatscale.atlassian.net/browse/ET-104?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/stores/cluster.tsx [hunk 1]
```diff
@@ -132,18 +132,20 @@ class ClusterStore extends PollingStore {
...
-          .filter((resource) => resource.enabled)
-          .forEach((resource) => {
-            const isResourceFree = resource.container == null;
-            const availableResource = isResourceFree ? 1 : 0;
-            overview[resource.type].available += availableResource;
-            overview[resource.type].total++;
-            overview[ResourceType.ALL].available += availableResource;
-            overview[ResourceType.ALL].total++;
-          });
+        agent.slotStats?.disabledSlots.length;
```
Comment (hamidzr) [inline webui/react/src/stores/cluster.tsx:136]: unused?

File: webui/react/src/types.ts [hunk 2]
```diff
@@ -241,6 +241,7 @@ export interface Agent {
   registeredTime: number;
   resourcePools: string[];
   slots?: SlotsRecord;
+  slotStats?: Api.V1SlotStats;
```
Comment (hamidzr) [inline webui/react/src/types.ts:244]: should always be available. but might be empty

**Review by hamidzr — APPROVED** [review]
lgtm we want to make sure agent.slots is not used anywhere in the cluster page but the topology view.

File: webui/react/src/stores/cluster.tsx (hunk 1 above)
Comment (keita-determined) [reply webui/react/src/stores/cluster.tsx:136]: oh yes. removed

File: webui/react/src/types.ts (hunk 2 above)
Comment (keita-determined) [reply webui/react/src/types.ts:244]: @hamidzr https://github.com/determined-ai/determined/pull/9048/files#diff-674d4524dc8c6e0f5971b4fcd5d6b220a2a5e05bbcf55954f9faf7883136cc1bR83, `slot_stats` should be added in `required` in protobuf because `slotStats` can be `undefined` now [code](https://github.com/determined-ai/determined/pull/9048/files#diff-58691f26d68f9a72c8a29a86d16b6cdb90e2719ce1e34b745bab5b6073b044fcR1098). will comment the same thing in your PR just in case

File: webui/react/src/stores/cluster.tsx [hunk 3]
```diff
@@ -132,18 +139,22 @@ class ClusterStore extends PollingStore {
...
+        const TYPE_PREFIX = 'TYPE_';
+        const types = Object.keys(agent.slotStats?.typeStats ?? {}).flatMap((key) => {
+          const type = key.replace(TYPE_PREFIX, '');
+          return isResourceType(type) ? [type] : [];
+        });
+        for (const type of types) {
+          const typeStat = agent.slotStats?.typeStats[`${TYPE_PREFIX}${type}`];
+          const total = typeStat?.total ?? 0;
+          const available =
+            total - Object.values(typeStat?.states ?? {}).reduce((a, b) => a + b, 0);
```
Comment (keita-determined) [inline webui/react/src/stores/cluster.tsx:152]: @hamidzr is it the expected way to calculate available slots?

Comment (hamidzr) [reply webui/react/src/stores/cluster.tsx:152]: I think so.

File: webui/react/src/services/apiConfig.ts [hunk 4]
```diff
@@ -481,7 +481,8 @@ export const getTelemetry: DetApi<EmptyParams, Api.V1GetTelemetryResponse, Telem
 export const getAgents: DetApi<EmptyParams, Api.V1GetAgentsResponse, Type.Agent[]> = {
   name: 'getAgents',
   postProcess: (response) => decoder.jsonToAgents(response.agents || []),
-  request: () => detApi.Cluster.getAgents(),
+  request: () =>
+    detApi.Cluster.getAgents(undefined, undefined, undefined, undefined, undefined, true),
```
Comment (hamidzr) [inline webui/react/src/services/apiConfig.ts:485]: I have some changes in this branch where I was working toward parameterizing getting slots so that we can still use it for topology.

that branch also includes some changes for disabling topology

basically if and only if  agents < 1000 && slotCount < 10000 then we want to pull down slots and render toplogy

File: webui/react/src/types.ts (hunk 2 above)
Comment (keita-determined) [reply webui/react/src/types.ts]: fixed

**Comment by hamidzr** [comment]
before change
Screenshot 2024-03-26 at 21 31 40
after change
Screenshot 2024-03-26 at 21 32 00

File: webui/react/src/pages/ResourcePool/Topology.tsx [hunk 5]
```diff
@@ -59,6 +59,19 @@ const NodeElement: React.FC<PropsWithChildren<NodeElementProps>> = ({ name, slot
 };
 
 const Topology: React.FC<PropsWithChildren<Props>> = ({ nodes }) => {
+  const MAX_USABLE_NODES = 100;
+  if (nodes.length > MAX_USABLE_NODES) {
```
Comment (hamidzr) [inline webui/react/src/pages/ResourcePool/Topology.tsx:63]: we want to base this off of both max nodes max slots
how can we get access to max slots?

File: webui/react/src/pages/ResourcePool/Topology.tsx [hunk 6]
```diff
@@ -59,6 +59,19 @@ const NodeElement: React.FC<PropsWithChildren<NodeElementProps>> = ({ name, slot
...
+    return (
+      <Section title="Topology">
+        <div className={css.warningContainer}>
+          <p>
+            Current Topology view is disabled for pools with more than {MAX_USABLE_NODES} nodes.
```
Comment (hamidzr) [inline webui/react/src/pages/ResourcePool/Topology.tsx:68]: I think the updated suggestion was that we just hide this @hkumar92

File: webui/react/src/stores/cluster.tsx [hunk 7]
```diff
@@ -206,10 +231,19 @@ class ClusterStore extends PollingStore {
 
   public async poll() {
     const agentRequest = getAgents({}, { signal: this.canceler?.signal });
+    const agentWithSlotsRequest = getAgents(
+      { excludeSlots: true },
```
Comment (hamidzr) [inline webui/react/src/stores/cluster.tsx:235]: shouldn't this be false?

Comment (hamidzr) [reply webui/react/src/stores/cluster.tsx:235]: question: I can't tell if we're doing this or not but the point is check and to not request slots when we have many agents and or slots otherwise web hits the same perf issues.

have you been able to test with the scaled up agents to see the effect?

**Comment by hamidzr** [comment]
thanks for working on this! I posted some questions and comments

File: webui/react/src/pages/ResourcePool/Topology.tsx (hunk 5 above)
Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/Topology.tsx]: yes. should be fixed

File: webui/react/src/pages/ResourcePool/Topology.tsx (hunk 6 above)
Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/Topology.tsx]: yep should be fixed

File: webui/react/src/stores/cluster.tsx (hunk 7 above)
Comment (keita-determined) [reply webui/react/src/stores/cluster.tsx]: right, should be false. yeah this should be handled now

File: webui/react/src/pages/ResourcePool/Topology.tsx (hunk 5 above)
Comment (hkumar92) [reply webui/react/src/pages/ResourcePool/Topology.tsx]: @keita-determined did we agree on 1,000 nodes instead of 100?

Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/Topology.tsx]: Yes. the latest code condition is `nodes > 1,000 OR slots > 10,000`

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx [hunk 8]
```diff
@@ -62,32 +63,70 @@ const ResourcepoolDetailInner: React.FC = () => {
...
   const agents = Loadable.getOrElse([], useObservable(clusterStore.agents));
   const resourcePools = useObservable(clusterStore.resourcePools);
+  const navigate = useNavigate();
+  const [canceler] = useState(new AbortController());
+
+  const [tabKey, setTabKey] = useState<TabType>(tab ?? DEFAULT_POOL_TAB_KEY);
+  const [poolsStats, setPoolsStats] = useState<V1RPQueueStat[]>();
+
+  const MAX_USABLE_NODES = 1_000;
+  const MAX_USABLE_SLOTS = 10_000;
```
Comment (ashtonG) [inline webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx:73]: lift this out of the closure

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx [hunk 9]
```diff
@@ -62,32 +63,70 @@ const ResourcepoolDetailInner: React.FC = () => {
...
 
+  const totalSlots = useMemo(() => {
+    const total = agents
+      .map((agent) => {
+        let totalSlots = 0;
+        for (const val of Object.values(agent.slotStats.typeStats ?? {})) {
+          totalSlots += val.total;
+        }
+        return totalSlots;
+      })
```
Comment (ashtonG) [inline webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx:89]: perhaps?
```suggestion
      .map((agent) =>
        Object.values(agent.slotStats.typeStats ?? {}).reduce((totalSlots, { total }) => totalSlots + total, 0)
      )
```

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx [hunk 10]
```diff
@@ -62,32 +63,70 @@ const ResourcepoolDetailInner: React.FC = () => {
...
+    return total;
+  }, [agents]);
+
+  const agentsWithSlots = useAsync(async () => {
+    if (agents.length > MAX_USABLE_NODES || totalSlots > MAX_USABLE_SLOTS) {
+      return [];
+    }
+    try {
+      const response = await getAgents({ excludeSlots: false });
+      return response;
```
Comment (ashtonG) [inline webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx:100]: should be able to just do
```suggestion
      return getAgents({ excludeSlots: false });
```

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx [hunk 11]
```diff
@@ -62,32 +63,70 @@ const ResourcepoolDetailInner: React.FC = () => {
...
+    } catch (e) {
+      handleError(e, { publicSubject: 'Could not agents with slots' });
```
Comment (ashtonG) [inline webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx:102]:
```suggestion
      handleError(e, { publicSubject: 'Could not get agents with slots' });
```

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx [hunk 12]
```diff
@@ -277,7 +316,20 @@ const ResourcepoolDetailInner: React.FC = () => {
             size={ShirtSize.Large}
           />
         </Section>
-        {!!topologyAgentPool.length && poolname && <Topology nodes={topologyAgentPool} />}
+        {agents.length > MAX_USABLE_NODES || totalSlots > MAX_USABLE_SLOTS ? (
+          <Section title="Topology">
+            <div className={css.warningContainer}>
+              <p>
+                Current Topology view is disabled for pools with more than {MAX_USABLE_NODES} nodes
```
Comment (hamidzr) [inline webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx:323]: HK was suggesting we do not show it instead of a msg here

File: webui/react/src/utils/cluster.ts [hunk 13]
```diff
@@ -1,20 +1,34 @@
...
+    .filter((agent) => {
+      if (resourcePoolName === undefined) {
+        return true;
+      }
+      for (const agentResourcePool of agent.resourcePools) {
+        if (!resourcePoolName.includes(agentResourcePool)) {
+          return false;
+        }
+      }
+      return true;
```
Comment (ashtonG) [inline webui/react/src/utils/cluster.ts:18]: can we keep this filter the same as it was before? the `!resourcePoolName.includes(agentResourcePool) -> false` thing is kind of hard to follow and not equivalent to the previous `agentResourcePools?.includes(resourcePoolName)`.

File: webui/react/src/utils/cluster.ts [hunk 14]
```diff
@@ -1,20 +1,34 @@
...
     })
-    .reduce((acc, resource) => [...acc, ...resource], [])
-    .filter((resource) => resource.enabled && resource.container)
-    .map((resource) => resource.container?.state) as ResourceState[];
-
+    .flatMap((agent) => {
+      const arr: ResourceState[] = Object.entries(agent.slotStats.typeStats ?? {})
+        .filter(([type]) => {
+          return resourceType === ResourceType.ALL ? true : type === `TYPE_${resourceType}`;
```
Comment (ashtonG) [inline webui/react/src/utils/cluster.ts:23]:
```suggestion
          return resourceType === ResourceType.ALL || type === `TYPE_${resourceType}`;
```

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx (hunk 8 above)
Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx]: fixed

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx (hunk 9 above)
Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx]: ah forgot to refactor code here. 

i end up doing this. i think i can reduce 1 loop with this

```ts
const total = agents.reduce(
      (totalVal, { slotStats }) =>
        totalVal +
        Object.values(slotStats.typeStats ?? {}).reduce(
          (localTotal, { total }) => localTotal + total,
          0,
        ),
      0,
    );
```

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx (hunk 10 above)
Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx]: fixed

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx (hunk 11 above)
Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx]: fixed

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx (hunk 12 above)
Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx]: ok when condition is matched, nothing shows up

**Review by ashtonG — COMMENTED** [review]
javascript looks good

**Comment by hamidzr** [comment]
I can see the toplogy turning on and off as expected. but I also do see a request to fetch all the slots in a scenario I wasn't expecting
Screenshot 2024-03-27 at 15 39 54

this case has 1m slots

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx [hunk 15]
```diff
@@ -52,6 +53,10 @@ type TabType = ValueOf<typeof TabType>;
 
 export const DEFAULT_POOL_TAB_KEY = TabType.Active;
 
+// PERF: hide topology if nodes or slots are huge to avoid rendering issue
+const MAX_USABLE_NODES = 1_000;
+const MAX_USABLE_SLOTS = 10_000;
```
Comment (hamidzr) [inline webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx:58]: I didn't know this is a thing 🎉

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx [hunk 16]
```diff
@@ -277,7 +316,11 @@ const ResourcepoolDetailInner: React.FC = () => {
             size={ShirtSize.Large}
           />
         </Section>
-        {!!topologyAgentPool.length && poolname && <Topology nodes={topologyAgentPool} />}
+        {agents.length <= MAX_USABLE_NODES && totalSlots <= MAX_USABLE_SLOTS && (
```
Comment (hamidzr) [inline webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx:319]: I'd probably share the check here w/ thee one that's deciding to load agents w/ slots so that they don't go out of sync

**Review by ashtonG — APPROVED** [review]

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx (hunk 15 above)
Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx:58]: which thing? :)

Comment (hamidzr) [reply webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx:58]: defining numbers like this 10_000

File: webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx (hunk 16 above)
Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx]: will fix it

Comment (keita-determined) [reply webui/react/src/pages/ResourcePool/ResourcepoolDetail.tsx]: fixed

**Review by hamidzr — CHANGES_REQUESTED** [review]
thanks for the update. I can see it pulling down the slots fine now when it should. But also we'd want it to keep refreshing those and continue to keep requesting for them. I assume that's needed for the topology to show the latest data
Screenshot 2024-03-27 at 21 36 51

**Comment by keita-determined** [comment]
> thanks for fixing that issue. I see the ui is double requesting agents now on a recurring basis we can say we want the latest numbers on agents to be able to turn topology on or off if the situation changes but maybe just relying on the initial agent response once is enough?
> this only happens when the conditions to show the topology is met eg the cluster is smaller so it's probably fine?

i think this should be fine because both APIs are used for different purposes and calling APIs is cheaper than the original issue.

**Comment by hamidzr** [comment]
test steps for posterity
https://github.com/determined-ai/release-party-issues/issues/1658#issuecomment-2035167365
